clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -66,143 +67,206 @@ class device_vector {
+      for (size_type i = 0; i < n; i++) {
+        std::allocator_traits<Allocator>::construct(_alloc,
+                                                    _storage + start_idx + i);
+      }
+    }
+  }
+
+  void _construct(size_type n, const T &value, size_type start_idx = 0) {
+    if (n > 0) {
+      for (size_type i = 0; i < n; i++) {
+        std::allocator_traits<Allocator>::construct(
+            _alloc, _storage + start_idx + i, value);
+      }
+    }
+  }
+
+  template <typename Iter>
+  void _construct(Iter first, Iter last, size_type start_idx = 0) {
+    size_type num_eles = ::std::distance(first, last);
+    if (num_eles > 0) {
+      size_type i = 0;
+      for (auto iter = first; iter != last; iter++, i++) {
+        std::allocator_traits<Allocator>::construct(
+            _alloc, _storage + start_idx + i, *iter);
+      }
+    }
+  }
+
+  void _destroy(size_type n, size_type start_idx = 0) {
+    if (n > 0) {
+      for (size_type i = 0; i < n; i++) {
+        std::allocator_traits<Allocator>::destroy(_alloc,
+                                                  _storage + start_idx + i);
+      }
+    }
   }
 
 public:
   template <typename OtherA> operator ::std::vector<T, OtherA>() const {
     auto __tmp = ::std::vector<T, OtherA>(this->size());
     ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-              this->begin(), this->end(), __tmp.begin());
+                this->begin(), this->end(), __tmp.begin());
     return __tmp;
   }
-  device_vector()
-      : _alloc(get_default_queue()), _size(0), _capacity(_min_capacity()) {
+
+  device_vector(const Allocator &alloc)
+      : _alloc(alloc), _size(0), _capacity(_min_capacity()) {
     _set_capacity_and_alloc();
   }
-  ~device_vector() /*= default*/ { _alloc.deallocate(_storage, _capacity); };
-  explicit device_vector(size_type n) : device_vector(n, T()) {}
-  explicit device_vector(size_type n, const T &value)
-      : _alloc(get_default_queue()), _size(n) {
+
+  device_vector() : device_vector(get_default_queue()) {}
+
+  ~device_vector() /*= default*/ {
+    clear();
+    ::std::allocator_traits<Allocator>::deallocate(_alloc, _storage, _capacity);
+  }
+
+  explicit device_vector(size_type n, const Allocator &alloc)
+      : _alloc(alloc), _size(n) {
     _set_capacity_and_alloc();
-    if (_size > 0) {
-      ::std::fill(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                begin(), end(), T(value));
-    }
+    _construct(n);
   }
-  device_vector(const device_vector &other) : _alloc(get_default_queue()) {
-    _size = other.size();
-    _capacity = other.capacity();
-    _storage = _alloc.allocate(_capacity);
-    ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-              other.begin(), other.end(), begin());
+
+  explicit device_vector(size_type n) : device_vector(n, get_default_queue()) {}
+
+  explicit device_vector(size_type n, const T &value, const Allocator &alloc)
```

Comment:
Can we remove the need for the separate constructor in line 146 by having the constructor's signature be `explicit device_vector(size_type n, const T &value, const Allocator &alloc = get_default_queue())`

Reply:
Yes, I fixed these and a few others, using `const Allocator &alloc = Allocator(get_default_queue())`.